src/components/core/GeoChart.tsx

```diff
@@ -116,19 +144,60 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, IGeoCh
     }
 
     public renderVisualization() {
-        const classes: string = `gd-geo-component s-gd-geo-component flex-direction-${this.getFlexDirection()}`;
+        const { height } = this.props;
+        if (height !== undefined) {
+            return this.renderVisualizationContent(undefined, height);
+        }
+
         return (
-            <div className={classes}>
-                {this.renderLegend()}
+            <Measure client={true}>
+                {({ measureRef, contentRect }: MeasuredComponentProps) => {
+                    const { client: contentRectClient } = contentRect;
+                    return this.renderVisualizationContent(
+                        measureRef,
+                        contentRectClient && contentRectClient.height,
+                    );
+                }}
+            </Measure>
+        );
+    }
+
+    private renderVisualizationContent(
+        measureRef: MeasuredComponentProps["measureRef"],
+        height: number,
+    ): JSX.Element {
+        const {
+            config: { legend: { responsive = false } = {} },
+        } = this.props;
+        const { showFluidLegend } = this.state;
+
+        const position = this.getLegendPosition(height);
+
+        const flexDirection = this.getFlexDirection(position);
+        const classes: string = cx("viz-line-family-chart-wrap", "gd-geo-component", "s-gd-geo-component", {
+            "responsive-legend": responsive,
+            "non-responsive-legend": !responsive,
+            [`flex-direction-${flexDirection}`]: true,
+            "legend-position-bottom": position === BOTTOM,
+        });
+        const isLegendRenderedFirst: boolean = position === TOP || (position === LEFT && !showFluidLegend);
+
+        return (
+            <div className={classes} ref={measureRef}>
+                {isLegendRenderedFirst && this.renderLegend(position)}
                 {this.renderChart()}
+                {!isLegendRenderedFirst && this.renderLegend(position)}
```

Comment:
I suggest separate to
```suggestion
                {this.renderTopLegend()}
                {this.renderChart()}
                {this.renderBottomLegend()}
```

Reply:
It is not about the position (top/bottom). 
It is about the render order, legend is rendered before/after chart.
Create 2 new different functions for that purpose is not necessary :)